src/comlink.ts

```diff
@@ -441,24 +445,32 @@ interface FinalizationRegistry<T> {
   ): void;
   unregister(unregisterToken: object): void;
 }
-declare var FinalizationRegistry: FinalizationRegistry<Endpoint>;
 
-const proxyCounter = new WeakMap<Endpoint, number>();
+declare var FinalizationRegistry: FinalizationRegistry<EndpointWithPendingListeners>;
+
+const proxyCounter = new WeakMap<EndpointWithPendingListeners, number>();
 const proxyFinalizers =
   "FinalizationRegistry" in globalThis &&
-  new FinalizationRegistry((ep: Endpoint) => {
-    const newCount = (proxyCounter.get(ep) || 0) - 1;
-    proxyCounter.set(ep, newCount);
-    if (newCount === 0) {
-      releaseEndpoint(ep);
+  new FinalizationRegistry(
+    (epWithPendingListeners: EndpointWithPendingListeners) => {
+      const newCount = (proxyCounter.get(epWithPendingListeners) || 0) - 1;
+      proxyCounter.set(epWithPendingListeners, newCount);
+      if (newCount === 0) {
+        releaseEndpoint(epWithPendingListeners).finally(() => {
+          epWithPendingListeners.pendingListeners.clear();
+        });
+      }
     }
-  });
+  );
 
-function registerProxy(proxy: object, ep: Endpoint) {
-  const newCount = (proxyCounter.get(ep) || 0) + 1;
-  proxyCounter.set(ep, newCount);
+function registerProxy(
+  proxy: object,
+  epWithPendingListeners: EndpointWithPendingListeners
+) {
+  const newCount = (proxyCounter.get(epWithPendingListeners) || 0) + 1;
```

Comment:
```suggestion
  const newCount = (proxyCounter.get(epWithPendingListeners) ?? 0) + 1;
```

Reply:
I'm leaving this as is as the logical OR operator works with much older browser versions. The nullish coalescing operator was introduced much later.

Reply:
fwiw - that is handled during the transpilation step of ts -> js. When you specify a specific ES version as the target it will produce the code for that ES version so you can write your ts using modern styles.